src/services/mapCode.ts

```diff
@@ -0,0 +1,329 @@
+import {
+    Block,
+    ClassElement,
+    ClassLikeDeclaration,
+    createSourceFile,
+    emptyArray,
+    FileTextChanges,
+    find,
+    findAncestor,
+    findLast,
+    flatten,
+    forEach,
+    formatting,
+    getTokenAtPosition,
+    isBlock,
+    isClassElement,
+    isClassLike,
+    isForInOrOfStatement,
+    isForStatement,
+    isIfStatement,
+    isInterfaceDeclaration,
+    isLabeledStatement,
+    isNamedDeclaration,
+    isSourceFile,
+    isTypeElement,
+    isWhileStatement,
+    LanguageServiceHost,
+    Mutable,
+    Node,
+    NodeArray,
+    or,
+    ScriptTarget,
+    some,
+    SourceFile,
+    Statement,
+    SyntaxKind,
+    textChanges,
+    TextSpan,
+    TypeElement,
+    UserPreferences,
+} from "./_namespaces/ts.js";
+import { ChangeTracker } from "./textChanges.js";
+
+/** @internal */
+export function mapCode(
+    sourceFile: SourceFile,
+    contents: string[],
+    focusLocations: TextSpan[][] | undefined,
+    host: LanguageServiceHost,
+    formatContext: formatting.FormatContext,
+    preferences: UserPreferences,
+): FileTextChanges[] {
+    try {
+        return textChanges.ChangeTracker.with(
+            { host, formatContext, preferences },
+            changeTracker => {
+                const parsed = contents.map(parse);
+                const flattenedLocations = focusLocations && flatten(focusLocations);
+                for (const nodes of parsed) {
+                    placeNodeGroup(
+                        sourceFile,
+                        changeTracker,
+                        nodes,
+                        flattenedLocations,
+                    );
+                }
+            },
+        );
+    }
+    catch (e) {
+        host.error?.(`mapCode: ${e}`);
+        return emptyArray;
+    }
+}
+
+/**
+ * Tries to parse something into either "top-level" statements, or into blocks
+ * of class-context definitions.
+ */
+function parse(content: string): NodeArray<Node> {
+    // We're going to speculatively parse different kinds of contexts to see
+    // which one makes the most sense, and grab the NodeArray from there. Do
+    // this as lazily as possible.
+    const nodeKinds = [
+        {
+            parse: () =>
+                createSourceFile(
+                    "__mapcode_content_nodes.ts",

```

Comment:
Doesnt this need to have same extension as sourceFile passed in as well as scriptTarget same as sourceFile  to ensure they are compatible

Reply:
nope. The node erasure we do later makes sure this isn't an issue.

Reply:
Hmm wierd it just seems not right to be to use random target and ".ts" extension to parse a file.. it feels like it should use same scanner and parser as input sourceFile's target and language version